// react-native-navigation/lib/dist/commands/Deprecations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deprecations = void 0;
const tslib_1 = require("tslib");
const once_1 = tslib_1.__importDefault(require("lodash/once"));
const get_1 = tslib_1.__importDefault(require("lodash/get"));
const each_1 = tslib_1.__importDefault(require("lodash/each"));
const react_native_1 = require("react-native");
class Deprecations {
    constructor() {
        this.deprecatedOptions = [
            {
                key: 'topBar.searchBarHiddenWhenScrolling',
                showWarning: once_1.default((_key, parentOptions) => {
                    console.warn(`${_key} is deprecated and will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/issues/6585`, parentOptions);
                }),
            },
            {
                key: 'topBar.searchBarPlaceholder',
                showWarning: once_1.default((_key, parentOptions) => {
                    console.warn(`${_key} is deprecated and will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/issues/6585`, parentOptions);
                }),
            },
            {
                key: 'topBar.searchBarBackgroundColor',
                showWarning: once_1.default((_key, parentOptions) => {
                    console.warn(`${_key} is deprecated and will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/issues/6585`, parentOptions);
                }),
            },
            {
                key: 'topBar.searchBarTintColor',
                showWarning: once_1.default((_key, parentOptions) => {
                    console.warn(`${_key} is deprecated and will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/issues/6585`, parentOptions);
                }),
            },
            {
                key: 'topBar.hideNavBarOnFocusSearchBar',
                showWarning: once_1.default((_key, parentOptions) => {
                    console.warn(`${_key} is deprecated and will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/issues/6585`, parentOptions);
                }),
            },
        ];
        this.deprecateSearchBarOptions = once_1.default((parentOptions) => {
            console.warn(`toggling searchBar visibility using a boolean value will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/issues/6585`, parentOptions);
        });
        this.deprecateInterpolationOptions = once_1.default((parentOptions) => {
            console.warn(`Using Interpolation types as strings has been deprecated and will be removed in the next major version. For more information see https://github.com/wix/react-native-navigation/pull/6644`, parentOptions);
        });
        this.deprecateBottomTabsVisibility = once_1.default((parentOptions) => {
            console.warn(`toggling bottomTabs visibility is deprecated on iOS. For more information see https://github.com/wix/react-native-navigation/issues/6416`, parentOptions);
        });
    }
    checkForDeprecatedOptions(options) {
        each_1.default(this.deprecatedOptions, (option) => {
            if (get_1.default(options, option.key, null)) {
                option.showWarning(option.key, options);
            }
        });
    }
    onProcessOptions(key, parentOptions, commandName) {
        if (key === 'bottomTabs' &&
            parentOptions[key].visible !== undefined &&
            react_native_1.Platform.OS === 'ios' &&
            commandName === 'mergeOptions') {
            this.deprecateBottomTabsVisibility(parentOptions);
        }
        if (key === 'searchBar' && react_native_1.Platform.OS === 'ios' && typeof parentOptions[key] === 'boolean') {
            this.deprecateSearchBarOptions(parentOptions);
        }
        if (key === 'interpolation' && typeof parentOptions[key] === 'string') {
            this.deprecateInterpolationOptions(parentOptions);
        }
    }
    onProcessDefaultOptions(_key, _parentOptions) { }
}
exports.Deprecations = Deprecations;
